Add checklist generation script (#80)
wordpress-mobile/release-toolkit-gutenberg-mobile: templates/release_checklist.md

<!-- Checklist generated by release toolkit -->

<!-- scheduled_release_only -->

> ## Pre release

Reviewer: In the original checklist, this section is named `Before the release`, I'm wondering if we should stick with that name, wdyt?

Reviewer: Not sure if it's doable but it would be great if we could add the date related to this part.

Author: Yep, I had the same thought about adding a date here. I guess in an updated flow we would run this checklist script on the first Monday of the release that way we can use the date the script was ran.

Reviewer: > Yep, I had the same thought about adding a date here. I guess in an updated flow we would run this checklist script on the first Monday of the release that way we can use the date the script was ran.

That would be a great enhancement 👍.

> - [ ] Clone or update the [release toolkit](https://href.li/?https://github.com/wordpress-mobile/release-toolkit-gutenberg-mobile) and review the [release script instructions](https://github.com/wordpress-mobile/release-toolkit-gutenberg-mobile/blob/develop/Releasing.md).

Reviewer: I noticed that this step is done in the `Create the release` section in the original checklist, right before running the release automation script. What do you think about moving the step there?

Author: I found it odd that it was in the "Create the release" section since it's mostly a prep step. For someone new it would be better to review the process a few days before. Or if it's been awhile since wrangling a release.
I suppose there could be an important change to the release script between, say Tuesday and Thursday but I doubt that would happen. In that case we could have the release automation script check for updated when it gets ran on Thursday

Reviewer: Good point, let's keep it here as a prep step for the release 👍 .

- [ ] [Notifiy open Gutenberg Mobile PRs assigned to {{version_number}} milestone]({{milestone_url}}).
<details>
<summary>Message Template</summary>
<p>

```markdown
Hello!
We will cut the {{version_number}} release on {{release_date}}.
I plan to circle back and bump this PR to the next milestone then,
but please let me know if you’d rather us work to include this PR in {{version_number}}.
Thanks!
```

</p>
</details>

> - [ ] Create WPiOS and WPAndroid builds from `develop`

Reviewer: ```suggestion
- [ ] Create WPiOS and WPAndroid builds from `develop`.
```

- [ ] [Perform general writing flow tests on Android](https://github.com/wordpress-mobile/test-cases/tree/master/test-cases/gutenberg/writing-flow).
- [ ] [Perform general writing flow tests on iOS](https://github.com/wordpress-mobile/test-cases/tree/master/test-cases/gutenberg/writing-flow).

<!-- /scheduled_release_only -->

> ## Create the Release

Reviewer: Following the original checklist, what do you think about adding the date in the heading?

```suggestion
## Create the Release ({{release_date}})
```

Author: I actually went back and forth on that. I had it there for awhile and then felt it was a bit redundant. I'm good with adding it back.

> ## Create the Release ({{release_date}})

Reviewer: Since we don't have a release date in beta and hotfixes, we need to have a different heading depending on the release type:

```suggestion
<!-- scheduled_release_only -->
## Create the Release ({{release_date}})
<!-- /scheduled_release_only -->
<!-- non_scheduled_release_only -->
## Create the Release
<!-- /non_scheduled_release_only -->
```

<!-- scheduled_release_only -->

> On {{release_date}}:

Reviewer: Since we already have the release date in the heading I think we could remove this line.

<!-- /scheduled_release_only -->

- [ ] [Verify Aztec is up to date](https://github.com/wordpress-mobile/release-toolkit-gutenberg-mobile/blob/develop/Releasing.md#handling-aztec-updates).

> - [ ] Run `./release_automation.sh` from the release toolkit to generate the main apps integration PRs [\*](#).

Reviewer: ```suggestion
- [ ] Run `./release_automation.sh` from the release toolkit to generate the main apps integration PRs.
```

Reviewer: I think it would be helpful to explicitly say to review the [release script instructions](https://github.com/wordpress-mobile/release-toolkit-gutenberg-mobile/blob/develop/Releasing.md) because depending on the release type, it's important to set the correct branches before running the script.

- [ ] Notify team that the release has been cut.
<details>
<summary>Message Template</summary>
<p>

```markdown
⚠️ The gutenberg-mobile {{version_number}} release branches are now cut.
Please do not merge any Gutenberg-related changes into the WPiOS or WPAndroid
develop branches until after the main apps cut their own releases next week.
If you’d like to merge changes now, merge them into the `gutenberg/after_{{version_number}}` branches.
```

</p>

> </details>

Reviewer: I think this step is only required for scheduled releases, right? Maybe we could wrap it with `<!-- scheduled_release_only -->`.

> - [ ] [Verify localization string files have been generated properly](https://github.com/wordpress-mobile/release-toolkit-gutenberg-mobile/blob/develop/Releasing.md).

Reviewer: ```suggestion
- [ ] [Verify localization string files have been generated properly](https://github.com/wordpress-mobile/release-toolkit-gutenberg-mobile/blob/develop/Releasing.md#verify-localization-strings).
```

> - [ ] Update `Unreleased` section in [`RELEASE-NOTES.txt`](https://github.com/wordpress-mobile/gutenberg-mobile/blob/develop/RELEASE-NOTES.txt) and [`gutenberg/../CHANGELOG.md`](https://github.com/WordPress/gutenberg/blob/trunk/packages/react-native-editor/CHANGELOG.md). Add a new `Unreleased` section after updating the release notes.

Reviewer: ```suggestion
- [ ] Update `Unreleased` section in [`RELEASE-NOTES.txt`](https://github.com/wordpress-mobile/gutenberg-mobile/blob/develop/RELEASE-NOTES.txt) and [`gutenberg/../CHANGELOG.md`](https://github.com/WordPress/gutenberg/blob/trunk/packages/react-native-editor/CHANGELOG.md) with the release version. Add a new `Unreleased` section after updating the release notes.
```

> - [ ] Perform quick smoke test of the editor on both iOS and Android builds

Reviewer: ```suggestion
- [ ] Perform quick smoke test of the editor on both iOS and Android builds to verify it launches without crashing.
```

Reviewer: I think it would be helpful to add the following extra steps that would help new members when running their first release:

> - Verify the WPAndroid PR build succeeds. If PR CI tasks include a 403 error related to an inability to resolve the `react-native-bridge` dependency, you must wait for the `Build Android RN Bridge & Publish to S3` task to succeed in gutenberg-mobile and then restart the WPAndroid CI tasks.

> - Trigger an installable build on WPiOS PR.

Reviewer: On the other hand, for beta and hotfixes probably this step is not necessary as we test the release before integrating it.

- [ ] Update [Gutenberg Mobile PR](https://github.com/wordpress-mobile/gutenberg-mobile/pulls?q=is%3Apr+is%3Aopen+label%3Arelease-process) description. Make sure the "Changes" section informs testers on what is in the release.

> - [ ] Mark all 4 PRs ready for review.

Reviewer: ```suggestion
- [ ] Mark all 4 PRs ready for review and request review from release wranglers.
```

<!-- non_scheduled_release_only -->

> - [ ] Notify the team members directly resposible for handling the releases of WPiOS and WPAndroid.

Reviewer: It would be great if we include a message template for this, it could be something like:

> Hi team. I wanted to provide a heads up that the Gutenberg Mobile team is starting work on the {{version_number}} beta release into the {{main_apps_version_number}} release branches. We'll provide an update once the integration is complete. Let us know if you have any questions or concerns. Thanks!

**NOTE:** For this message we would need to ask the version number of the apps.

<!-- /non_scheduled_release_only -->

> <!-- optional_aztec_release -->

Reviewer: I'm wondering if we should move this section to after the `Verify Aztec is up to date` step, as this process requires pushing changes to both Gutenberg Mobile and Gutenberg release branches and would be better to do it before setting the PRs ready for review, wdyt?

> <!-- optional_incoming_changes -->

Reviewer: I think the incoming changes are only expected in scheduled releases so we could wrap this with `<!-- scheduled_release_only -->`, wdyt?

> ## Integrate the Release

Reviewer: For consistency, I'd add the release date in this heading too as we have in [this line](https://github.com/wordpress-mobile/release-toolkit-gutenberg-mobile/pull/80/files#diff-4bd888d653a69f531e87b9da8d17c2bb2ee6343316f0ef6bee965327e5f8c684R30).

<!-- scheduled_release_only -->

> On {{release_date}}:

Reviewer: Following the previous comment, I'd remove this line if we add the release date to the heading of the section.

<!-- /scheduled_release_only -->

> - [ ] Verify `gutenberg` ref within the gutenberg-mobile release branch is pointing to the [latest commit](#) in Gutenberg release branch.

Reviewer: ```suggestion
- [ ] Verify `gutenberg` ref within the gutenberg-mobile release branch is pointing to the latest commit in Gutenberg release branch.
```

- [ ] Create and push a `rnmobile/{{version_number}}` git tag for the head of [Gutenberg release branch](https://github.com/WordPress/gutenberg/branches/all?query=rnmobile%2Frelease).

> - [ ] [Update bundles in the Gutenberg Mobile release branch](#).

Reviewer: As per the corresponding step from the original checklist, I don't think is mandatory to update the bundle files, so maybe we should update the message, wdyt?

> Ensure that the bundle files are updated to include any changes to the release branch by running npm run bundle in gutenberg-mobile release branch and committing any changes.

> - [ ] Merge **gutenberg-mobile** PR to `trunk` [using the "Create a merge commit" option](#). **WARNING** do **not** merge gutenberg PR to `trunk` yet.

Reviewer: ```suggestion
- [ ] Merge **gutenberg-mobile** PR to `trunk` using the "Create a merge commit" option. **WARNING** do **not** merge gutenberg PR to `trunk` yet.
```

> - [ ] [Create a new Gutenberg moble release](#).

Reviewer: I think we could generate this link in the script as it only requires the new version name.

**Example:** `https://github.com/wordpress-mobile/gutenberg-mobile/releases/new?tag=v1.65.0&target=trunk&title=Release%201.65.0`

> - [ ] [Create a new Gutenberg moble release](#).

Reviewer: ```suggestion
- [ ] [Create a new Gutenberg mobile release](#).
```

> - [ ] [Update Gutenberg mobile references in the main apps](#).

Reviewer: I'm wondering if you were considering adding a section in the release documentation for this, if not, I'd consider adding the following steps here:

> In WPiOS, update the reference to point to the tag of the Release created in the previous task.

> In WPAndroid, update the `gutenbergMobileVersion` in `build.gradle` to point to the tag of the Release used in the previous task.

Author: I was going to add this extra info to the release documentation. I prefer to have the checklist item as a reminder and  the details available via a link if some needs to learn or be reminded on how to update the main apps.

Reviewer: Ok, I think it's a great idea to include the details of these steps in the release documentation, the clearer the steps are the better to avoid mistakes 😅 .

> I was going to add this extra info to the release documentation.

Are you planning to add them in this PR or we should remove the link in the meantime?

- [ ] Re-run optional tests on main apps PRs.

> - [ ] Merge main apps PSs to the respective `develop` branches.

Reviewer: ```suggestion
- [ ] Merge main apps PRs to the respective `develop` branches.
```

Reviewer: We should change this step depending on the release type, as for beta and hotfixes, the target is the release branch (e.g. `release/17.9`).

> - [ ] Notify team that the release is now in the main apps `develop` branches

Reviewer: ```suggestion
- [ ] Notify team that the release is now in the main apps `develop` branches.
```

Reviewer: Maybe we should specify which team should be notified with this message, wdyt?

Reviewer: Same as in my previous message, for beta and hotfixes the target is the release branch (e.g. release/17.9).

<details>
<summary>Message Template</summary>
<p>

```markdown
Hey team. I wanted to let you know that the mobile Gutenberg team has finished integrating
the {{version_number}} Gutenberg release into the WPiOS and WPAndroid `develop` branches.
The integration is ready for the next release cut/build creation when you are available.
Please let me know if you have any questions. Thanks!
```

</p>

> </details>

Reviewer: Depending on the release type, we should provide a different message:

### Scheduled
> Hey team. I wanted to let you know that the mobile Gutenberg team has finished integrating the X.XX.X Gutenberg release into the WPiOS and WPAndroid `develop` branches. The integration is ready for the next release cut/build creation when you are available. Please let me know if you have any questions. Thanks! 

### Betafix/hotfix
> Hey team. I wanted to let you know that the mobile Gutenberg team has finished integrating the X.XX.X Gutenberg release into the WPiOS and WPAndroid `release/XX.X` branches, ready for a new beta when you are available. Please let me know if you have any questions. Thanks!

## Sync Release

> - [ ] Merge release changes back to `gutenberg-mobile/develop` e.g. `trunk -> merge_release_{{version_number}}_to_develop -> develop`

Reviewer: After checking the original checklist, I think this step should be placed in the second position:

> 1. If there are any conflicts in the gutenberg PR, merge `trunk` into it and resolve them.

> 2. In gutenberg-mobile, create a branch from `trunk` with a name like `merge_release_1.65.0_to_develop` and open PR to `develop`. If there are any merge conflicts, merge `develop` into the PR and resolve them.

> - [ ] Reslove any conflicts and merge `gutenberg/rnmobile/release_{{release_version}}` PR to `gutenberg/trunk`.

Reviewer: ```suggestion
- [ ] Resolve any conflicts and merge `gutenberg/rnmobile/release_{{version_number}}` PR to `gutenberg/trunk`.
```

- [ ] Update `gutenberg` reference in `gutenberg-mobile/merge_release_{{version_number}}_to_develop` to point to `gutenberg/trunk`.
- [ ] Merge `gutenberg-mobile/merge_release_{{version_number}}_to_develop` PR to `gutenberg/develop`.

## Clean Up Pending Work

⚠️ This section may only be completed after the main apps cut their own release branches.

- [ ] Open a PR to merge `gutenberg/after_{{version_number}}` into `gutenberg-mobile/develop`.

> - [ ] Merge or Close`gutenberg/after_{{version_number}}` **AFTER** the main apps have cut release branches.

Reviewer: ```suggestion
- [ ] Merge or close `gutenberg/after_{{version_number}}` only **AFTER** the main apps have cut release branches.
```

## Test the Release

Use the main WP apps to complete the following tests:

#### iOS

> - [ ] Test new [changes in the release](https://github.com/wordpress-mobile/gutenberg-mobile/blob/HEAD/RELEASE-NOTES.txt)

Reviewer: ```suggestion
- [ ] Test new [changes in the release](https://github.com/wordpress-mobile/gutenberg-mobile/blob/HEAD/RELEASE-NOTES.txt).
```

> - [ ] Complete the [general writing flow tests].(https://github.com/wordpress-mobile/test-cases/tree/HEAD/test-cases/gutenberg/writing-flow

Reviewer: ```suggestion
- [ ] Complete the [general writing flow tests](https://github.com/wordpress-mobile/test-cases/tree/HEAD/test-cases/gutenberg/writing-flow).
```

- [ ] Complete the [Unsupported Block Editor test cases](https://github.com/wordpress-mobile/test-cases/blob/HEAD/test-cases/gutenberg/unsupported-block-editing.md#unsupported-block-editing---test-cases).
- [ ] Complete the [Sanity Tests](https://github.com/wordpress-mobile/test-cases/blob/HEAD/test-suites/gutenberg/sanity-test-suites.md)

#### Android

> - [ ] Test new [changes in the release](https://github.com/wordpress-mobile/gutenberg-mobile/blob/develop/RELEASE-NOTES.txt)

Reviewer: ```suggestion
- [ ] Test new [changes in the release](https://github.com/wordpress-mobile/gutenberg-mobile/blob/develop/RELEASE-NOTES.txt).
```

> - [ ] Complete the [general writing flow tests].(https://github.com/wordpress-mobile/test-cases/tree/HEAD/test-cases/gutenberg/writing-flow

Reviewer: ```suggestion
- [ ] Complete the [general writing flow tests](https://github.com/wordpress-mobile/test-cases/tree/HEAD/test-cases/gutenberg/writing-flow).
```

<!-- Checklist generated by release toolkit -->

<!-- scheduled_release_only -->

## Before the Release ({{before_release_date}})

- [ ] Review the [release documentation](https://github.com/wordpress-mobile/release-toolkit-gutenberg-mobile/blob/develop/Releasing.md).
- [ ] [Notifiy open Gutenberg Mobile PRs assigned to {{gb_mobile_version}} milestone]({{milestone_url}}).
<details>
<summary>Message Template</summary>
<p>

```markdown
Hello!
We will cut the {{gb_mobile_version}} release on {{release_date}}.
I plan to circle back and bump this PR to the next milestone then,
but please let me know if you’d rather us work to include this PR in {{gb_mobile_version}}.
Thanks!
```

</p>
</details>

> - [ ] Clone or update the [release toolkit](https://github.com/wordpress-mobile/release-toolkit-gutenberg-mobile)

Reviewer: Looks like we need to separate these lines with an empty line otherwise, the markdown is not properly rendered (see attached screenshot).

<img width="500" alt="Screenshot 2021-11-11 at 10 43 43" src="https://user-images.githubusercontent.com/14905380/141275846-679aca5a-a0ad-4eab-a4f3-a330317340e4.png">

```suggestion
</details>

- [ ] Clone or update the [release toolkit](https://github.com/wordpress-mobile/release-toolkit-gutenberg-mobile).
```

- [ ] Create WPiOS and WPAndroid builds from `develop`.
- [ ] [Perform general writing flow tests on Android](https://github.com/wordpress-mobile/test-cases/tree/master/test-cases/gutenberg/writing-flow).
- [ ] [Perform general writing flow tests on iOS](https://github.com/wordpress-mobile/test-cases/tree/master/test-cases/gutenberg/writing-flow).

<!-- /scheduled_release_only -->

## Create the Release ({{release_date}})

- [ ] [Verify Aztec is up to date](https://github.com/wordpress-mobile/release-toolkit-gutenberg-mobile/blob/develop/Releasing.md#handling-aztec-updates).
- [ ] Run `./release_automation.sh` from the release toolkit to generate the main apps integration PRs [\*](#).

> - [ ] Notify team that the release has been cut.

Reviewer: I miss the step for bumping the opened PRs assigned to the milestone, we already notified the authors in the pre release section but I think it would be helpful to explicitly add this step, wdyt?

**Example:**
```
- [ ] [Bump open Gutenberg Mobile PRs assigned to {{gb_mobile_version}} milestone]({{milestone_url}}) to next milestone.
```

<details>
<summary>Message Template</summary>
<p>

```markdown
⚠️ The gutenberg-mobile {{gb_mobile_version}} release branches are now cut.
Please do not merge any Gutenberg-related changes into the WPiOS or WPAndroid
develop branches until after the main apps cut their own releases next week.
If you’d like to merge changes now, merge them into the `gutenberg/after_{{gb_mobile_version}}` branches.

> ```

Reviewer: This message is related to scheduled releases, so we should provide a different message or omit the template, wdyt?

</p>
</details>

- [ ] [Verify localization string files have been generated properly](https://github.com/wordpress-mobile/release-toolkit-gutenberg-mobile/blob/develop/Releasing.md).
- [ ] Update `Unreleased` section in [`RELEASE-NOTES.txt`](https://github.com/wordpress-mobile/gutenberg-mobile/blob/develop/RELEASE-NOTES.txt) and [`gutenberg/../CHANGELOG.md`](https://github.com/WordPress/gutenberg/blob/trunk/packages/react-native-editor/CHANGELOG.md). Add a new `Unreleased` section after updating the release notes.
- [ ] Perform quick smoke test of the editor on both iOS and Android builds
- [ ] Update [Gutenberg Mobile PR](https://github.com/wordpress-mobile/gutenberg-mobile/pulls?q=is%3Apr+is%3Aopen+label%3Arelease-process) description. Make sure the "Changes" section informs testers on what is in the release.
- [ ] Mark all 4 PRs ready for review.

<!-- non_scheduled_release_only -->

> - [ ] Notify the team members directly resposible for handling the releases of WPiOS and WPAndroid.

Reviewer: ```suggestion
- [ ] Notify the team members directly responsible for handling the releases of WPiOS and WPAndroid.
```

## Test the Release

Use the main WP apps to complete the following tests:

#### iOS

- [ ] Test new [changes in the release](https://github.com/wordpress-mobile/gutenberg-mobile/blob/HEAD/RELEASE-NOTES.txt).
- [ ] Complete the [general writing flow tests](https://github.com/wordpress-mobile/test-cases/tree/HEAD/test-cases/gutenberg/writing-flow).
- [ ] Complete the [Unsupported Block Editor test cases](https://github.com/wordpress-mobile/test-cases/blob/HEAD/test-cases/gutenberg/unsupported-block-editing.md#unsupported-block-editing---test-cases).
- [ ] Complete the [Sanity Tests](https://github.com/wordpress-mobile/test-cases/blob/HEAD/test-suites/gutenberg/sanity-test-suites.md)

#### Android

- [ ] Test new [changes in the release](https://github.com/wordpress-mobile/gutenberg-mobile/blob/develop/RELEASE-NOTES.txt).
- [ ] Complete the [general writing flow tests](https://github.com/wordpress-mobile/test-cases/tree/HEAD/test-cases/gutenberg/writing-flow).
- [ ] Complete the [Unsupported Block Editor test cases](https://github.com/wordpress-mobile/test-cases/blob/HEAD/test-cases/gutenberg/unsupported-block-editing.md#unsupported-block-editing---test-cases).

> - [ ] Complete the [Sanity Tests](https://href.li/?https://github.com/wordpress-mobile/test-cases/blob/HEAD/test-suites/gutenberg/sanity-test-suites.md)

Reviewer: Usually, we don't cover all these tests on beta and hotfixes, I'm wondering we could provide a reduced list for those cases. Besides, tests for beta and hotfixes should be done before integrating the release, so we could consider moving this before the `Integrate the Release` section only for those release types.
